server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/BrowseHelper.java

```diff
@@ -0,0 +1,99 @@
+import java.time.Clock;
+import java.time.Instant;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+
+import org.apache.james.queue.api.ManageableMailQueue;
+import org.apache.james.queue.rabbitmq.MailQueueName;
+import org.apache.james.queue.rabbitmq.view.cassandra.model.EnqueuedMail;
+import org.apache.james.util.FluentFutureStream;
+
+class BrowseHelper {
+
+    private final BrowseStartDAO browseStartDao;
+    private final DeletedMailsDAO deletedMailsDao;
+    private final EnqueuedMailsDAO enqueuedMailsDao;
+    private final CassandraMailQueueViewConfiguration configuration;
+    private final Clock clock;
+
+    @Inject
+    BrowseHelper(BrowseStartDAO browseStartDao,
+                 DeletedMailsDAO deletedMailsDao,
+                 EnqueuedMailsDAO enqueuedMailsDao,
+                 CassandraMailQueueViewConfiguration configuration, Clock clock) {
+        this.browseStartDao = browseStartDao;
+        this.deletedMailsDao = deletedMailsDao;
+        this.enqueuedMailsDao = enqueuedMailsDao;
+        this.configuration = configuration;
+        this.clock = clock;
+    }
+
+    CompletableFuture<Stream<ManageableMailQueue.MailQueueItemView>> browse(MailQueueName queueName) {
+        return browseReferences(queueName)
+            .map(EnqueuedMail::getMail)
+            .map(ManageableMailQueue.MailQueueItemView::new)
+            .completableFuture();
+    }
+
+    FluentFutureStream<EnqueuedMail> browseReferences(MailQueueName queueName) {
+        return FluentFutureStream.of(browseStartDao.findBrowseStart(queueName)
+            .thenApply(this::allSlicesFrom))
+            .thenFlatMap(slice -> browseSlice(queueName, slice));
+    }
+
+    private FluentFutureStream<EnqueuedMail> browseSlice(MailQueueName queueName, Slice slice) {
+        return FluentFutureStream.ofFluentFutureStreams(
+            allBucketIds()
+                .map(bucketId -> browseBucket(queueName, slice, bucketId)))
+            .sorted(Comparator.comparing(EnqueuedMail::getEnqueuedTime));
+    }
+
+    private FluentFutureStream<EnqueuedMail> browseBucket(MailQueueName queueName, Slice slice, BucketId bucketId) {
+        return FluentFutureStream.of(
+            enqueuedMailsDao.selectEnqueuedMails(queueName, slice, bucketId))
+                .thenFilter(mailReference -> deletedMailsDao.isStillEnqueued(queueName, mailReference.getMailKey()));
+    }
+
+    private Stream<Slice> allSlicesFrom(Optional<Instant> maybeBrowseStart) {
+        return maybeBrowseStart
+            .map(browseStart -> Slice.of(browseStart, configuration.getSliceWindow()))
+            .map(startSlice -> allSlicesTill(startSlice, clock.instant()))
+            .orElse(Stream.empty());
+    }
+
+    private Stream<BucketId> allBucketIds() {
+        return IntStream
+            .range(0, configuration.getBucketCount())
```

Comment:
any configuration change in the number of bucket will break the browsing ? how are we going to handle that ?

Reply:
You can safely increase the number of buckets, not decrease it.

This IMO is OK as it allows to **scale out** but not **scale down**.